### PR TITLE
Add more overloads to LoggerMessage.Define

### PR DESCRIPTION
**Pull request by pakrym**
@lodejard

### CODE REVIEW

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs [hunk 1]
```diff
@@ -69,6 +69,26 @@ public static class LoggerMessage
         /// <summary>
         /// Creates a delegate which can be invoked for logging a message.
         /// </summary>
+        /// <param name="logLevel">The <see cref="LogLevel"/></param>
+        /// <param name="eventId">The event id</param>
+        /// <param name="formatString">The named format string</param>
+        /// <returns>A delegate which when invoked creates a log message.</returns>
+        public static Action<ILogger, Exception> Define(LogLevel logLevel, int eventId, string formatString)
```
Comment (kichalla) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: This extension should not be required as the intention of the extensions in this file is to optimize for the scenarios where we pass parameters to the format string (where as here we do not pass any)...

Example:
```
public static void OrderPlaced(this ILogger logger)
{
    if (logger.IsEnabled(LogLevel.Information)
    { 
        logger.LogInformation(eventId: 10, "Order placed");
    }
}
```

Comment (pakrym) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: True, but we get LoggingExtensions with two usage patterns inside which doesn't look good.

Comment (Eilon) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: @kichalla  I said it was ok :smile:

**Comment by Eilon** [comment]
This addresses this issue: https://github.com/aspnet/Logging/issues/272

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs (hunk 1 above)
Comment (Eilon) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: @kichalla in this particular case it's more about consistency and not any actual perf concern.

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs [hunk 2]
```diff
@@ -268,6 +363,125 @@ public LogValues(LogValuesFormatter formatter, T0 value0, T1 value1, T2 value2,
...
+                new KeyValuePair<string, object>(_formatter.ValueNames[5], _value5),
+                new KeyValuePair<string, object>(_formatter.ValueNames[6], _value6),
+                new KeyValuePair<string, object>("{OriginalFormat}", _formatter.OriginalFormat),
+            };
+
+            public object[] ToArray() => new object[] { _value0, _value1, _value2, _value3, _value4, _value5, _value6 };
+
+            public override string ToString() => _formatter.Format(ToArray());
+        }
+
```
Comment (Eilon) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: :open_mouth: extra lines!

**Comment by Eilon** [comment]
Is it possible to add at least a basic test for these?

File: src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs (hunk 1 above)
Comment (kichalla) [inline src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs]: Sound good...I agree.

**Comment by pakrym** [comment]
@eilon will do

**Comment by pakrym** [comment]
:up: :date:

File: test/Microsoft.Extensions.Logging.Test/LoggerMessageTest.cs [hunk 3]
```diff
@@ -162,6 +163,48 @@ public void LogScope_WithThreeParameters()
                 actualLogValues.ToString());
         }
 
+        [Theory]
+        [MemberData(nameof(LogMessagesData))]
+
```
Comment (Eilon) [inline test/Microsoft.Extensions.Logging.Test/LoggerMessageTest.cs]: Remove space :trollface:

File: test/Microsoft.Extensions.Logging.Test/LoggerMessageTest.cs [hunk 4]
```diff
@@ -162,6 +163,48 @@ public void LogScope_WithThreeParameters()
...
+            var testSink = new TestSink();
+            var testLogger = new TestLogger("testlogger", testSink, enabled: true);
+            var exception = new Exception("TestException");
+            var parameterNames = Enumerable.Range(0, argumentCount).Select(i => "P" + i).ToArray();
+            var parameters = new List<object>();
+            parameters.Add(testLogger);
+            parameters.AddRange(parameterNames);
+            parameters.Add(exception);
+
+            var expectedFormat = "Log " + string.Join(" ", parameterNames.Select(p=>"{"+p+"}"));
```
Comment (Eilon) [inline test/Microsoft.Extensions.Logging.Test/LoggerMessageTest.cs]: Format codes :trollface:

**Comment by Eilon** [comment]
:shipit: after code formatting.

**Pull request closed without merging**